src/hotspot/share/nmt/indexedFreeListAllocator.hpp

```diff
@@ -0,0 +1,102 @@
+/*
+ * Copyright (c) 2024, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+
+ */
+
+#include "utilities/growableArray.hpp"
+
+#ifndef SHARE_NMT_INDEXEDFREELISTALLOCATOR_HPP
+#define SHARE_NMT_INDEXEDFREELISTALLOCATOR_HPP
+
+// A free list, growth only, allocator for a specific type E.
+// The allocator returns 'pointers' of 4-bytes in size, allowing for
+// memory savings if a pointer-heavy self-referential structure is used.
+// It is "indexed" as a reference is base + index * sizeof(E).
+// It never returns any memory to the system.
+template<typename E, MEMFLAGS flag>
+class IndexedFreeListAllocator {
+public:
+  using I = int32_t;
```

Comment:
This is what I'd like to have in a follow-up RFE. Your last two points are because of `GrowableArray`: GA uses int32, so we do, and GA checks overflow when expanding, so we're OK.